test/distributed/fsdp/test_fsdp.py

```diff
@@ -0,0 +1,252 @@
+        tgt = self.embed_tokens(tgt_ids)
+        tgt = self.bn(tgt)
+        x = self.transformer(src, tgt)
+        return self.output_proj(x)
+
+    def get_loss(self, input, output):
+        _, tgt = input
+        return nn.functional.cross_entropy(output.view(-1, output.size(-1)), tgt.view(-1), reduction="sum")
+
+    def run_backward(self, loss):
+        loss.backward()
+
+
+class NestedWrappedModule(nn.Module):
+    def __init__(self, group, wrap_fsdp, wrap_everything=False):
+        super().__init__()
+        self.rank = group.rank()
+        self.world_size = group.size()
+
+        def _maybe_wrap(layer):
+            if wrap_fsdp:
+                return FullyShardedDataParallel(layer, group)
+            return layer
+
+        torch.manual_seed(0)  # keep everything deterministic
+        self.module = nn.Sequential(
+            nn.Linear(8, 4),
+            _maybe_wrap(nn.Sequential(_maybe_wrap(nn.Linear(4, 16)), nn.Linear(16, 16),)),
+            _maybe_wrap(nn.Linear(16, 4)),
+            nn.Linear(4, 8),
+        )
+
+        if wrap_everything:
+            self.module = nn.Sequential(
+                _maybe_wrap(nn.Linear(8, 4)),
+                _maybe_wrap(nn.Linear(4, 16)),
+                _maybe_wrap(nn.Linear(16, 4)),
+                _maybe_wrap(nn.Linear(4, 8)),
+            )
+
+    def get_input(self, device):
+        torch.manual_seed(1 + self.rank)  # keep everything deterministic
+        return (torch.rand(4, 8, device=device),)
+
+    def forward(self, x):
+        return self.module(x)
+
+    def get_loss(self, input, output):
+        loss = output.sum()
+        return loss
+
+    def run_backward(self, loss):
+        loss.backward()
+
+
+class DistributedTest(MultiProcessTestCase):
+    def setUp(self):
+        super(DistributedTest, self).setUp()
+        self._spawn_processes()
+
+    @property
+    def world_size(self):
+        return torch.cuda.device_count() if torch.cuda.is_available() else 4
+
+    @property
+    def init_method(self):
+        return "{}{file_name}".format(FILE_SCHEMA, file_name=self.file_name)
+
+    @classmethod
+    def _run(cls, rank, test_name, file_name, pipe):
+        self = cls(test_name)
+        self.rank = rank
+        self.file_name = file_name
+
+        print(f"dist init r={self.rank}, world={self.world_size}")
+        os.environ["WORLD_SIZE"] = str(self.world_size)
+        os.environ["RANK"] = str(self.rank)
+
+        backend = "nccl" if torch.cuda.is_available() else "gloo"
+        try:
+            dist.init_process_group(
+                init_method=self.init_method,
+                backend=backend,
+                world_size=int(self.world_size),
+                rank=self.rank,
+            )
+        except RuntimeError as e:
+            if "recompile" in e.args[0]:
+                sys.exit(TEST_SKIPS["backend_unavailable"].exit_code)
+
+            raise
```

Comment:
These looks a bit different from our existing distributed tests where they didn't manually set the env var. Since we already explicitly passing `rank` and `world_size` to `init_process_group`, do we still need to set the env vars?

https://github.com/pytorch/pytorch/blob/c7ef620a14b1cadf0e36d637dba9226b919ac3ff/test/distributed/test_c10d_gloo.py#L1457-L1459

Reply:
feel it is more easier to run test without depending on environment variables. probably we do not need to set the env var...